design/host-profiles.md

```diff
@@ -0,0 +1,2008 @@
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+         * [Example Ironic Inventory Results](#example-ironic-inventory-results)
+            * [Virtual Machine](#virtual-machine)
+            * [Bare Metal](#bare-metal)
+
+<!-- Added by: stack, at: 2019-02-15T16:58-05:00 -->
+
+<!--te-->
+
+## Summary
+
+Host profiles allow us to match the hardware characteristics of a host
+to provide a label similar to an instance flavor in a traditional
+cloud environment. Host profiles are configurable through the
+Kubernetes API, allowing cluster administrators to expand the known
+hardware platforms and control how they are used in the cluster.
+
+## Motivation
+
+We want to limit the provisioning operations to known hardware
+platforms, while making the set of known platforms easy to expand over
+time. We do not want to hard-code the definitions, because that makes
+it more difficult to configure development or test environments that
+may not match what an expected production environment looks like.
+
+### Goals
+
+- Provide a place for us to store known hardware profiles.
+- Provide a way to match known profiles to actual hardware.
+- Provide a way to control how that hardware is used as part of a
+  MachineSet.
+
+### Non-Goals
+
+This goal is about examining the hardware present in the host. It does
+not consider whether that hardware is returning fault codes or
+otherwise not working properly. Dealing with faulty hardware is
+covered under a separate design.
+
+For now, we only want to do simple matching based on data, and we do
+not want a complex "language" to describe matching rules. It is enough
+for hosts to look like the hardware described in the profile.
+
+This design does not cover using the profile to control how software
+is deployed to the host (specifying networking access, storage use,
+etc.). That may come in a later design.
+
+## Proposal
+
+A new CRD, BareMetalHostProfile, is defined with Spec fields for all
+of the relevant hardware considerations. Ideally this would reuse the
+hardware specification data structure in the BareMetalHost itself, or
+be a different structure that summarizes that structure (for example,
+providing numCPUs and cpuGHz, fields but not a list of CPUs and their
+individual speeds).
+
+The Status portion of the BareMetalHostProfile will report all of the
+BareMetalHosts and MachineSets associated with the profile.
+
+*We need more detail about exactly what sort of hardware matching we
+care about.*
+
+When a profile is created, the profile operator causes all host
+resources labeled as having an unknown profile to be reconciled to see
+if they match the new profile. This means the order of creation for
+hosts and profiles does not matter, because the match will happen
+eventually.
+
+When a profile is updated, the profile operator causes all host
+resources associated with that profile to be reconciled to see if they
+still match the updated profile.
```

Comment:
Agreed on the above - but I also need to educate myself on exactly what constitutes a profile.  Today it isn't much.  For instance, one thing we struggle with outside of metal3-io is that lots of things can constitute a profile, even things like firmware and RAID configurations.  If you change any of these in the declaration, our intent is that the host is "dirty" and should be corrected at the next _opportune_ time, e.g. in this context, the next time it is associated with a Machine or if someone does something manually that explicitly acknowledges its OK to re-provision this machine.   While we want to avoid pets, its a large leap to say touching a common profile could result in a whole lot of Machine churn.

Reply:
The point of this step was just to see if the host state matches the profile rules. It was not to make the host match the profile. I don't think we want to do rely on changes to the profile causing reprovisioning or BIOS updates of hosts. Those should be controlled through the MachineDeployment settings, because a MachineDeployment knows how to roll the changes out across several machines safely.

Reply:
So what you're saying is that a profile serves as a way to keep inventory on hosts with particular configurations? If so, it makes sense that changing a profile would not cause changes to the host to be made, it would just cause that host to no longer have a match to any of the created profiles. 

If a host has its profile changed to unknown due to not matching exactly any one created profile, we should be sure to save in the status of the CRD the last profile it did match cuz it's useful to figure out how or why this host no longer matches its previous matched profile.